src/VisualStudio/Core/Def/Implementation/Options/RoamingVisualStudioProfileOptionPersisterProvider.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Settings;
+using IAsyncServiceProvider = Microsoft.VisualStudio.Shell.IAsyncServiceProvider;
+using SAsyncServiceProvider = Microsoft.VisualStudio.Shell.Interop.SAsyncServiceProvider;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Options
+{
+    [Export(typeof(IOptionPersisterProvider))]
+    internal sealed class RoamingVisualStudioProfileOptionPersisterProvider : IOptionPersisterProvider
+    {
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsyncServiceProvider _serviceProvider;
+        private readonly IGlobalOptionService _optionService;
+        private RoamingVisualStudioProfileOptionPersister? _lazyPersister;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RoamingVisualStudioProfileOptionPersisterProvider(
+            IThreadingContext threadingContext,
+            [Import(typeof(SAsyncServiceProvider))] IAsyncServiceProvider serviceProvider,
+            IGlobalOptionService optionService)
+        {
+            _threadingContext = threadingContext;
+            _serviceProvider = serviceProvider;
+            _optionService = optionService;
+        }
+
+        public IOptionPersister? TryGetPersister()
+        {
+            return _lazyPersister;
+        }
+
+        public async ValueTask<IOptionPersister> GetOrCreatePersisterAsync(CancellationToken cancellationToken)
+        {
+            if (_lazyPersister is not null)
+            {
+                return _lazyPersister;
+            }
+
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var settingsManager = (ISettingsManager?)await _serviceProvider.GetServiceAsync(typeof(SVsSettingsPersistenceManager)).ConfigureAwait(true);
+
+            _lazyPersister ??= new RoamingVisualStudioProfileOptionPersister(_threadingContext, _optionService, settingsManager);
+            return _lazyPersister;
+        }
```

Comment:
common pattern for all (except creation of the persister).  consider introducing an abstract base that shares all this.

Reply:
I don't want to force implementations to construct objects on the main thread. Between this and other refactorings to remove `TryGetPersister`, there isn't enough left to extract and still have a net reduction in complexity.